### PR TITLE
build both x86_64 and arm64 rsession components on macOS

### PR DESCRIPTION
**Pull request by kevinushey**
### Intent

This PR seeks to make it possible to use a single copy of RStudio with both x86_64 and arm64 builds of R on macOS.

### Approach

Roughly:

1. Produce a full build of RStudio for x86_64,
2. Produce an `rsession` build for arm64,
3. Copy that `rsession` (and its library dependencies) into the built tree for x86_64,
4. Build RStudio as a package with both binaries,
5. When RStudio is launched, choose the appropriate `rsession` based on the architecture of the underlying R library.

Note: I chose this approach over the "universal binary" approach as it's much easier to adapt our build scripts in this case. In particular, we depend on Homebrew for some libraries and Homebrew does not provide universal binaries or libraries in any case.

This also only effects package builds; developer builds will remain single-architecture. (This is mainly because I found getting CMake to build the same sub-directory in the same project with different compilers / compiler flags is rather challenging.)

The implementation here is also designed such that regular package builds on x86_64 should produce the exact same builds as we had before; any arm64 components should only be activated if we detect that the running machine is itself running on arm64.

### Automated Tests

None required.

### QA Notes

None required specifically for this PR, but smoke-testing on M1 macOS machines will be warranted.

### Checklist

- [ ] If this PR adds a new feature, or fixes a bug in a previously released version, it includes an entry in `NEWS.md` 
- [ ] If this PR adds or changes UI, the updated UI meets [accessibility standards](https://github.com/rstudio/rstudio/wiki/Accessibility)
- [ ] A reviewer is assigned to this PR (if unsure who to assign, check Area Owners list)
- [ ] This PR passes all local unit tests

### CODE REVIEW

**Comment by gtritchie** [comment]
Note that Electron has chosen the separate builds approach (instead of creating a universal binary), though they plan to create a universal binary as well (but say it will be HUGE).

https://www.electronjs.org/blog/apple-silicon

File: package/osx/CMakeLists.txt [hunk 1]
```diff
@@ -21,43 +21,74 @@ endif()
...
+      execute_process(
+         COMMAND R -s -f \"${CMAKE_CURRENT_SOURCE_DIR}/fix-arm64-library-paths.R\"
+         --args
+         \"\${CMAKE_INSTALL_PREFIX}/RStudio.app/Contents/Frameworks\"
+         \"\${CMAKE_INSTALL_PREFIX}/RStudio.app/Contents/Frameworks/arm64\"
+         COMMAND_ERROR_IS_FATAL ANY)
+
+   endif()
+   
+   # don't follow symlinks in GLOB_RECURSE
```
Comment (kevinushey) [inline package/osx/CMakeLists.txt:71]: Note that nothing below here has changed (this is all just re-formatting to use a consistent 3-space indent)

File: dependencies/common/install-boost [hunk 2]
```diff
@@ -26,7 +26,7 @@ PLATFORM=`uname`
 # vars
 BOOST_VERSION_NUMBER=1.69.0
 BOOST_VERSION=boost_1_69_0
-BOOST_DIR=/opt/rstudio-tools/boost/$BOOST_VERSION
+BOOST_DIR=/opt/rstudio-tools/$(uname -m)/boost/$BOOST_VERSION
```
Comment (kevinushey) [inline dependencies/common/install-boost:29]: Note that this change would require rebuilding dependencies on x86_64 machines. Do you think that's fine, or should we consider setting this only for arm64?

File: dependencies/osx/install-dependencies-osx-arch [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+# AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+#
+#
+
+set -eux
+
+# NOTE: the intention is for this script to be invoked as e.g.
+#
+#   arch -x86_64 ./install-dependencies-osx-arch
+#   arch -arm64  ./install-dependencies-osx-arch
```
Comment (jmcphers) [inline dependencies/osx/install-dependencies-osx-arch:23]: Is there a way to detect and/or fail the script quickly if it wasn't invoked correctly?

File: package/osx/fix-arm64-library-paths.R [hunk 4]
```diff
@@ -0,0 +1,47 @@
+
```
Comment (jmcphers) [inline package/osx/fix-arm64-library-paths.R:1]: Should this be an executable R script w/ shebang?

File: src/cpp/desktop/DesktopMain.cpp [hunk 5]
```diff
@@ -827,6 +828,39 @@ int main(int argc, char* argv[])
...
+         {
+            LOG_ERROR(error);
+         }
+
+         std::string rArchs = string_utils::trimWhitespace(result.stdOut);
+         if (rArchs.find("arm64") != std::string::npos)
+         {
+            FilePath arm64SessionPath = sessionPath.getParent().completeChildPath("rsession-arm64");
+            if (arm64SessionPath.exists())
+               sessionPath = arm64SessionPath;
```
Comment (jmcphers) [inline src/cpp/desktop/DesktopMain.cpp:858]: We should log a warning or error here as there is almost certainly trouble ahead.

**Review by jmcphers — COMMENTED** [review]
Very exciting, so glad this didn't turn out to be a big hassle (or, it appears, require any core code changes at all).

File: dependencies/common/install-boost (hunk 2 above)
Comment (jmcphers) [reply dependencies/common/install-boost:29]: I think it's fine. It does imply that we will need to rebuild deps manually on the build host, right?

**Comment by jmcphers** [comment]
As I read this, the build requirements for this change are as follows:

- After merging it, we will need to rebuild deps on the x86 build host before a build will be successful.
- The change does not cause us to start cross compiling for x86 on the arm64 host.
- The change will automatically produce both x86 and arm64 binaries on an arm64 host.

Does that sound right?

File: dependencies/osx/install-dependencies-osx-arch (hunk 3 above)
Comment (kevinushey) [reply dependencies/osx/install-dependencies-osx-arch:23]: Not really -- if not run through `arch`, then the install script just runs for whichever architecture is currently "active". This is what you want on x86_64 (where there's only one possible architecture); on arm64 this is useful when you want to ensure a particular script is run either natively or with the Rosetta translation layer active.

File: package/osx/fix-arm64-library-paths.R (hunk 4 above)
Comment (kevinushey) [reply package/osx/fix-arm64-library-paths.R:9]: Done!

**Comment by kevinushey** [comment]
That is correct, with the caveat that I think on your second point you meant "x86_64 host" :-)

Basically, none of our existing Jenkins build scripts should need to change, we should just "do the right thing" based on the architecture of the machine producing the build.

**Review by jmcphers — APPROVED** [review]
Very exciting, so glad this didn't turn out to be a big hassle (or, it appears, require any core code changes at all).

**Comment by jmcphers** [comment]
Let me know when you're ready to merge this and we can pair on getting the build host set up.

**Comment by kevinushey** [comment]
I decided to spend a bit more time to make sure that these changes don't compromise development workflows on macOS. (This is mainly recalling some of the pain we have with RStudio on Windows, where attempts to launch RStudio could fail if you hadn't told RStudio to use a 32-bit copy of R versus 64-bit copy of R).

To that end RStudio Server now respects `RSTUDIO_WHICH_R`, to make it easier to override the version of R used in development configurations. This is valid since one can build against x86_64 R and run against arm64 R (since at build time we only use the headers; it's only at runtime that we look up and load + insert the appropriate R library).

So now, if you're building RStudio on macOS and need to select an appropriate version of R after build, that becomes much easier and doesn't require re-compiling the project.

**Pull request merged**